exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java

```diff
@@ -587,4 +593,77 @@ private void checkStoppedService(Integer serviceId) {
   Optional<ServiceWrapper> findService(String name) {
     return Optional.ofNullable(services.get(name));
   }
+
+  /**
+   * Returns migration script for the given artifact to perform asynchronous data migration.
+   *
+   * @param artifactId Java service artifact id
+   * @param dataVersion base data version migrate from
+   * @return migration script instance or {@link Optional#empty()} if there is no scripts found
+   * @throws IllegalArgumentException if the provided artifact is not deployed
+   * @throws IllegalStateException if there is scripts incompatibility
+   */
+  public Optional<MigrationScript> migrate(ServiceArtifactId artifactId, String dataVersion) {
+    try {
+      synchronized (lock) {
+        LoadedServiceDefinition serviceDefinition = serviceLoader.findService(artifactId)
+            .orElseThrow(() -> new IllegalArgumentException(
+                "Migration called on unknown artifactId: " + artifactId));
+
+        if (serviceDefinition.getMigrationScripts().isEmpty()) {
+          logger.info("Service data migration is not required, skipping."
+              + " No scripts found for the given artifact {}", artifactId);
+          return Optional.empty();
+        }
+
+        Version artifactVersion = Version.valueOf(artifactId.getVersion());
+        Version baseDataVersion = Version.valueOf(dataVersion);
```

Comment:
If in `ServiceArtifactId` class change a type for `version` field then these transformations would be redundant. I don't insist actually. Maybe there are some valuable reasons to use `String` type for version 😄 I didn't dig deep, to be honest.

Reply:
There are the following points here worth to mention:
- Java <-> native integration requires to have base types (i.e. you should serialize/deserialize objects). So, String is a good choice for passing `Version` through.
- `ServiceArtifactId` is a public model object and to bring there `java-semver` dependency (and force users using it) is not so good idea I believe.
So, we use String and documented that versions should be in a valid _Semantic Versioning_ format.